File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int N;
    static int M;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] dy = {-1, 1, 0, 0};
        int[] dx = {0, 0, -1, 1};

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        int[][] grid = new int[N][M];
        boolean[][] visited = new boolean[N][M];

        for(int i = 0; i < N; i++){
            String[] arr = br.readLine().split("");
            for(int j = 0; j < M; j++){
                grid[i][j] = Integer.parseInt(arr[j]);
            }
        }

        visited[0][0] = true;

        Queue<int[]> q = new LinkedList<>();

        q.add(new int[]{0, 0, 1});

        while(!q.isEmpty()){
            int[] curr = q.poll();

            int y = curr[0];
            int x = curr[1];
            int count = curr[2];

            if(y == (N - 1) && x == (M - 1)){
                System.out.println(count);
                break;
            }

            for(int i = 0; i < 4; i++){
                int ny = y + dy[i];
                int nx = x + dx[i];

                if(0 <= ny && ny < N && 0 <= nx && nx < M && !visited[ny][nx] && grid[ny][nx] == 1){
                    q.add(new int[]{ny, nx, count + 1});
                    visited[ny][nx] = true;
                }
            }
        }
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

N, M = map(int, sys.stdin.readline().split())

graph = []
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

for _ in range(N):
    graph.append(list(map(int, sys.stdin.readline().strip())))

def bfs(a, b):
    q = deque([])
    q.append([a, b, 1])
    visited = [[False] * M for _ in range(N)]
    visited[a][b] = True

    while q:
        x, y, count = q.popleft()

        if x == N-1 and y == M-1:
            return count

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < N and 0 <= ny < M and graph[nx][ny] == 1 and not visited[nx][ny]:
                visited[nx][ny] = True
                q.append([nx, ny, count + 1])

print(bfs(0, 0))